src/core/stream/buffers_handler.ts

```diff
@@ -514,91 +517,145 @@ export default function BuffersHandler(
   ) : Observable<IPeriodBufferEvent> {
-    return adaptation$.pipe(switchMap((adaptation) => {
-      if (adaptation == null) {
-        log.info(`set no ${bufferType} Adaptation`, period);
-        let cleanBuffer$ : Observable<null>;
+    return adaptation$.pipe(
+      startWith(null),
+      pairwise(),
+      switchMap(([ oldAdaptation, adaptation ]: [Adaptation|null, Adaptation|null]) => {
+        if (adaptation == null) {
+          log.info(`set no ${bufferType} Adaptation`, period);
+          let cleanBuffer$ : Observable<null>;
+
+          if (sourceBufferManager.has(bufferType)) {
+            log.info(`clearing previous ${bufferType} SourceBuffer`);
+            const _queuedSourceBuffer = sourceBufferManager.get(bufferType);
+            cleanBuffer$ = _queuedSourceBuffer
+              .removeBuffer({ start: period.start, end: period.end || Infinity })
+              .pipe(mapTo(null));
+          } else {
+            cleanBuffer$ = observableOf(null);
+          }
+
+          return observableConcat(
+            cleanBuffer$.pipe(mapTo(EVENTS.adaptationChange(bufferType, null, period))),
+            createFakeBuffer(clock$, wantedBufferAhead$, bufferType, { manifest, period })
+          );
+        }
+
+        log.info(`updating ${bufferType} adaptation`, adaptation, period);
 
+        // 1 - create or reuse the SourceBuffer
+        let queuedSourceBuffer : QueuedSourceBuffer<any>;
         if (sourceBufferManager.has(bufferType)) {
-          log.info(`clearing previous ${bufferType} SourceBuffer`);
-          const _queuedSourceBuffer = sourceBufferManager.get(bufferType);
-          cleanBuffer$ = _queuedSourceBuffer
-            .removeBuffer({ start: period.start, end: period.end || Infinity })
-            .pipe(mapTo(null));
+          log.info("reusing a previous SourceBuffer for the type", bufferType);
+          queuedSourceBuffer = sourceBufferManager.get(bufferType);
         } else {
-          cleanBuffer$ = observableOf(null);
+          const codec = getFirstDeclaredMimeType(adaptation);
+          const sourceBufferOptions = bufferType === "text" ?
+            options.textTrackOptions : undefined;
+          queuedSourceBuffer = sourceBufferManager
+            .createSourceBuffer(bufferType, codec, sourceBufferOptions);
         }
 
+        // 2 - create or reuse the associated BufferGarbageCollector and
+        // SegmentBookkeeper
+        const bufferGarbageCollector$ = garbageCollectors.get(queuedSourceBuffer);
+        const segmentBookkeeper = segmentBookkeepers.get(queuedSourceBuffer);
+
+        // 3 - Clear from buffer all data from previous adaptation, in current period
+        const cleanBuffered$: Observable<void> =
+          (oldAdaptation != null) ? clock$.pipe(
+            take(1),
+            mergeMap(({ currentTime }) => {
+              const bufferedRanges = segmentBookkeeper
+                .getBufferedForAdaptation((oldAdaptation as Adaptation).id);
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this unnecessary cast. ([typescript:S4325](https://sonarcloud.io/organizations/rx-player/rules#rule_key=typescript%3AS4325))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=rx-player&pullRequest=278&issues=AWUaQY5S8anAXINB_PSu&open=AWUaQY5S8anAXINB_PSu)

[ISSUE_KEY]: AWUaQY5S8anAXINB_PSu